### ISSUE TITLE
.kdbx file in nextcloud don't open in keeweb

### ISSUE THREAD

**Issue by derjohannes19** [jhass/nextcloud-keeweb#34]
When I click on a .kdbx file the file will get downloaded, and it don't get open in keeweb.

Also: Webdav is not working, i get http 401 error.

**Comment by aalaesar** [comment]
+1
tested on nextcloud 11.0.0 & 10.0.2

**Comment by jackxb** [comment]
+1 
tested on nextcloud 11.0.0

**Comment by xamindar** [comment]
Same issue here. Just installed it in nextcloud 11 and nothing has changed regarding clicking on kdbx files. Click on them, and they are downloaded. They do not open in keeweb.

**Comment by geiseri** [comment]
Same here on 11.

**Comment by scottharwell** [comment]
I don't intend to add another "me too", but I have some other details.  I have found that if I disable the app on Nextcloud 11, and then restart it, then I am able to select a keepass db for about 30 mins.  After that, the webdav connection seems to "break" and keeweb seems to move the open DB to a managed db outside of the Nextcloud integration.  I then get the download behavior that others describe.  Deleting the connection from keeweb, disabling keeweb, enabling keeweb, and reopening the ldbx file then allows me to use it again.  Maybe this behaviour description will help.

**Comment by Echse27** [comment]
After install the KeeWeb App inside NC11.0.0, you can choose a .kdbx file only one time by klicking the file itself.
When you edit a entry or add one, there is no sync back to the main .kdbx database. 
Error: 401 
So now i cant found a way to drop the file from the WebApp...

**Comment by Yamakasi** [comment]
Opening in Keeweb worked for me but just stopped without updating anything, only my browser (Chromium) is updated.

**Comment by Huster-at-github** [comment]
Same here. Does it need to be registered as a mimetype?

**Comment by Yamakasi** [comment]
I'm not sure, when I had a browsercache issue when upgraded to NC11 it just worked like that after I fixed that (cleared more hours of cache) and now it just stopped.

I thought I needed Webdav so was kinda surprised that clicking on it just worked!

I hope @jhass that he can give us some idea.

**Comment by nschiwy** [comment]
I've now been able to "fix" this twice by disabling and then re-enabling the app. For the time being I'm probably just going to use a URL shortener to open the URL, which tells it the nextcloud location of the database. (ex. https://your-url.com/nextcloud/index.php/apps/keeweb/?open=%2Fpath%2Fto%2Fdatabase.kdbx) though I can't be sure whether this will allow it to save correctly...

**Comment by Yamakasi** [comment]
Re-enabeling is a very good tip, this works indeed and happened also at my side when I had caching issues and re-enabled the app after emptying my cache. Maybe some mimetype magic happens in the browsercache when you enable it ?

**Comment by rocky4546** [comment]
I have some info that may kinda help.  First, I found the command to update the database indexes that will cause files located in the data area to launch KeeWeb.  This command is
sudo -u www-data php [nextcloud home]/occ -vvv files:scan --all

I still have issues, but this is close.  If the file in the data area is update, either from KeeWeb or externally updated by KeePass, the launch breaks and the command needs to be run again to fix it.

If the file is shared across account, updating the file has been breaking it again and it has also been breaking the share.  The person that did the update is the only one with the update.  The other people don't get the update.

If the file is on an external storage, then the file may or may not work with KeeWeb.  In my case, I have 4 files in 2 folders.  One of the four work with KeeWeb, the others do not.  The occ command has no effect on the external storage even though the files are updated.

**Comment by juanjux** [comment]
As a tip, you don't need to disable and reenable the app to use the file from Keeweb: if you open the Keeweb app from the initial app menu the file will show in the list and opening it from there will work even if Keeweb doesn't open from the Files app.

**Comment by Yamakasi** [comment]
I will not always display there, I think this is due to sessions which are
not there anymore.

2016-12-31 20:26 GMT+01:00 juanjux <notifications@github.com>:

**Comment by dnauck** [comment]
For me the workaround with disable/enable the app or scan all files does not help.

I' ve the following message in my java script console:

    Resource interpreted as Document but transferred with MIME type application/octet-stream

Looks like the mime type is not registered.

**Comment by gadolithium** [comment]
+1 tested on NC 11.0.0.

**Comment by dennisheitmann** [comment]
+1 does not work (click on file in NC 11)

**Comment by juanjux** [comment]
@Yamakasi it's true, a little later I lost the session and it dissapeared from the "recent files" list. But maybe I found another workaround? Tested with several different browsers, computers, networks and incognito mode so the session shouldn't exist. If you go directly to the keeweb + file URL like:

https://mydomain.com/apps/keeweb/?open=/mifile.kdbx

It seems to work.

**Comment by grojas7** [comment]
+1
Tested on Nextcloud 11.0.0

**Comment by juanjux** [comment]
Yes, this is still happening on 11.0.1.

**Comment by SonireolxXx** [comment]
Same problem here on NC 11.0.1
I'm using WebDAV to sync with other devices, once the file is updated by other clients this problem occurs.

**Comment by mwalbeck** [comment]
I think I've nailed down the issue. When you edit and save a file, keeweb saves it in a new file with a unix timestamp after the name then proceeds to delete the old file and rename the new one to remove the timestamp. Nextcloud registers the file with the unix timestamp as the extension and caches that information, so the cached mimetype is wrong. 

Also the way the kdbx mimetype is registered, if you upload a keepass file after you've enabled the app it will be cached with a generic mimetype instead of the kdbx mimetype. I don't really know the details on this one.

I've created a small fix for the above. It's not the best solution but It works. It simply changes the mimetype of any .kdbx file in the filecache everytime you reload a page in nextcloud, so it might effect performance depending on your setup.

If you add

`$mimetypeId = $mimeTypeLoader->getId('x-application/kdbx');
$mimeTypeLoader->updateFilecache('%.kdbx', $mimetypeId);`

Below

`// Register custom mimetype we can hook in the frontend
$mimeTypeDetector->getAllMappings();
$mimeTypeDetector->registerType('kdbx', 'x-application/kdbx', 'x-application/kdbx');`

In your **appinfo/app.php** file you should be good. (It's at the bottom)

I can make a pull request if there is interest for this, but none the less I'll try to find a more elegant solution and make a pull request with that instead.

**Comment by SonireolxXx** [comment]
@mwalbeck  Thank you for the solution, I just tried it on NC 11.0.1 and it works, didn't notice any significant impact on performance.

I think this is the best approach for the issue so far.

**Comment by Yamakasi** [comment]
Magnus, good idea, create a PR for it I would say.

2017-01-18 22:37 GMT+01:00 SonireolxXx <notifications@github.com>:

**Comment by mwalbeck** [comment]
@jhass I've send a pull request to upstream nextcloud to fix the issue with mimetype detection on hidden files I talked about here #38. This is only half the problem though. From what I've found the mimetype detection doesn't utilize the oc_mimetypes database tables to look up mimetypes. So the only proper way to currently register the mimetype is in the mimetypemapping file in nextcloud server. So if the mimetype is registered with upstream nextcloud this issue would be resolved with the release of nextcloud 12.

Thoughts?

**Comment by jhass** [comment]
Well, registering the mimetype upstream would certainly work, but that feels like a band-aid to a missing feature in the app API, so I'd prefer to either have the detection use the database or have an API to ship a mimetype used for detection with an app, for example by shipping another json file that's aggregated with the default one.

**Comment by guddl** [comment]
I had this problem when I resticted the app to be available for a special group only. 
After removing the group from "Enable only for specific groups" everything works fine again.

**Comment by dedero1985** [comment]
Hi guys,

I tried the solution from mwalbeck and works great (thanks a lot!). I see no performance impacts under Nextcloud 11.0.1 using Keeweb v1.3.3 (52701af, 2016-12-17)

Cheers,
Bruno.

**Comment by ds4a** [comment]
The temporary fix offered by @nschiwy of 

> I've now been able to "fix" this twice by disabling and then re-enabling the app.

works for me also.

**Comment by Sanookmakmak** [comment]
Same here. Nextcloud 11.0.1 and Keeweb 0.3.0

Disabling and re-enabling the Keeweb app helps, but after a while it is no more possible to open the kdbx file. I just can be downloaded.

So the admin needs to disable and re-enable the app again (and again and again...)

:-(

**Comment by mwalbeck** [comment]
With the release of Nextcloud 12, you can fix this by adding a **mimetypemapping.json** file to your **nextcloud/config** folder. In the file you add the mimetype **"kdbx": ["x-application/kdbx"]**

For easy copy paste:

    {
        "kdbx": ["x-application/kdbx"]
    }

There is a little more info in the nextcloud admin docs here, https://docs.nextcloud.com/server/11/admin_manual/configuration_mimetypes/index.html

This will survive both app and system updates and should be the best solution until it can be done through the app itself.

**Comment by ds4a** [comment]
@mwalbeck I have read the documentation you referred to at https://docs.nextcloud.com/server/11/admin_manual/configuration_mimetypes/index.html.
I'm a little confused as this doesn't seem to present a problem to this issue. The documentation refers to the mimetypealiases and mimetypemapping json files that allow NC to display the correct icon for a specified mimetype. The icon files are placed in the relevant theme directory. Am I missing something?
I directly edited the mimetypealiases.dist.json and mimetypemapping.dist.json files located in resources/config in my NC installation and ran the maintenance:mimetype:update-js command, and the reported issue still persists.

**Comment by mwalbeck** [comment]
@ds4a I should probably have been a bit more specific, but if you only focus on the mimetypemapping aspect, which is a way to register a mimetype in nextcloud. Since this issue is a mimetype issue we can use the mimetypemapping file to register a mimetype in nextcloud. This is needed as there currently is no proper way for an app to register a new mimetype.

So we don't need to edit the mimetypealiases file or run the maintenance:mimetype:update-js command, as they are purely for matching icons and mimetypes.

Also, as I noted above, this will only work with Nextcloud 12 and onwards, because of a small change to the mimetype detection in Nextcloud.

**Comment by ds4a** [comment]
@mwalbeck Thanks for the explanation and clarification.

**Comment by Darkitty** [comment]
"I've now been able to "fix" this twice by disabling and then re-enabling the app."

Fix the issue with Nextcloud 11.0.1 and version 0.3.0 :D

**Comment by proehlen** [comment]
@mwalbeck 's code fix didn't fix the issue for me unfortunately.  Upgrading to NextCloud 12 also isn't an option.  I can't really disable/enable the app repeatedly either since the files change a lot.  Anyway got any other ideas on this?

**Comment by phalae** [comment]
The enable/disable "fix" worked for me, but How to also open the *.key file ? 💃 :)

**Comment by rocky4546** [comment]
I just checked and the update that mwalbeck suggested above works great has not been implemented. It forces the mimetype to be set correctly when you navigate to the folder and click on the .kdbx file. Here is a snippet of what he stated above.

(The fix) simply changes the mimetype of any .kdbx file in the filecache everytime you reload a page in nextcloud, so it might effect performance depending on your setup. If you add
$mimetypeId = $mimeTypeLoader->getId('x-application/kdbx');
$mimeTypeLoader->updateFilecache('%.kdbx', $mimetypeId);

Below
// Register custom mimetype we can hook in the frontend
$mimeTypeDetector->getAllMappings();
$mimeTypeDetector->registerType('kdbx', 'x-application/kdbx', 'x-application/kdbx');

In your keeweb/appinfo/app.php file you should be good. (It's at the bottom)

**Comment by fthiery** [comment]
@rocky4546 so the fix is to patch the keeweb/appinfo/app.php file that should look like the following, is that correct ?

```
// Register custom mimetype we can hook in the frontend
$mimeTypeDetector->getAllMappings();
$mimeTypeDetector->registerType('kdbx', 'x-application/kdbx', 'x-application/kdbx');
$mimetypeId = $mimeTypeLoader->getId('x-application/kdbx');
$mimeTypeLoader->updateFilecache('%.kdbx', $mimetypeId);
```
This is supposed to work on v11, right ?

**Comment by rocky4546** [comment]
That is what I implemented on two systems and each time it fixed the issue when the file was updated outside of keeweb.  I am running 11.0.0 and probably going to 11.0.2 shortly.  I really don't know why people are enabling and disabling the app to deal with this issue when such a nice work around is available.

**Comment by Darkitty** [comment]
Or, when I use the direct link with the file in URL, works perfect anytime without actions.

**Comment by grobz100** [comment]
I am also having this issue.

**Comment by riegercloud** [comment]
I had the same issue but disabling and re-enabling solved it as described above.

**Comment by 4oo4** [comment]
Like everyone else, whenever it stops working, the disable/re-enable fix seems to solve it. However I've also noticed that when you favorite your .kdbx file, that it breaks it and you have to apply the fix again.

**Comment by BetaLeaf** [comment]
Can we please get this pulled into MASTER? The fix was posted over 3 months ago and this still hasn't been done. https://apps.nextcloud.com/apps/keeweb

https://github.com/jhass/nextcloud-keeweb/issues/34#issuecomment-273595295

**Comment by jhass** [comment]
@BetaLeaf please have a look at the discussion in #38

**Comment by stratacast** [comment]
I just stumbled across this and decided to follow the disable/enable idea that people here have mentioned, and it worked for me too. One thing that was also interesting is I got a message that I had a message pop up in the app that said there was a major update to KeeWeb and to be sure to back up my db files. Perhaps we who disable/enable the app are pushing on an update that Nextcloud doesn't properly implement for whatever reason? I don't understand why 1 of my kdbx files would work but my others wouldn't unless I changed the filename in the URL. I'm using Nextcloud 11.0.3

**Comment by Eloo** [comment]
Just upgraded to NC12 and currently i'm unable to click and open my kdbx file.
Tried disable/enable and the mimetypemapping.json from @mwalbeck but withoud success.
Is any using NC12 with keeweb?

**Comment by Darkitty** [comment]
@Eloo Have you try to update KeeWeb ? A new version is out for support NC 12 : https://github.com/jhass/nextcloud-keeweb/commit/98eea5e13668825d8a5e02ceff0b23b37d60c182 :)

**Comment by Eloo** [comment]
Yep, latest version was 0.4.0?

**Comment by JSoko** [comment]
If anybody has this problem with NC12

nano /var/www/nextcloud/config/mimetypemapping.json

{
    "kdbx": ["x-application/kdbx"]
}

sudo -u www-data php /var/www/nextcloud/occ maintenance:mimetype:update-js

**Comment by geiseri** [comment]
in theory this will be fixed in NC12 by https://github.com/nextcloud/server/issues/4631.  Can anyone confirm this?  I do not have the beta here.

**Comment by JSoko** [comment]
It wasn't fixed for me with NC 12.0.0.29...

**Comment by mwalbeck** [comment]
@JSoko the actual command is

`maintenance:mimetype:update-db --repair-filecache`

the command you posted updates the mimetype references to icons which you can put in the mimetypealiases.json file to decide which mimetypes have which icons.

@geiseri sadly no, since the problem is the fact that nextcloud doesn't have a proper way for apps to register a mimetype and the fix you mentioned sadly doesn't change that fact.

**Comment by jthomasdewald** [comment]
@mwalbeck Thanks for the correct command. Worked on my NC 12.0.0 only after adding the mimetype, updating the db AND disabling/enabling keepass

**Comment by patschi** [comment]
Had the same reported issue as described here, with latest Nextcloud 12.0.0.29 (fresh install). Without any manually editing the `mimetypemapping.json`, just the `mimetype:update-db` command fixed it. I'm not that fimilar with the possibilities of the NC-plugin-development, but wouldn't it be possible to just trigger that command automatically when enabling this plugin?

**Comment by AgenturVS** [comment]
Thanks, juanjux. I'm using your URL and it is working perfect for me.

**Comment by stratacast** [comment]
I did what @JSoko suggested and ran the command that @mwalbeck said to do and that fixed my problem with opening my .kdbx file on NC 12.0.0 with keeweb version 0.4. Thank you!

**Comment by KingPrawn22** [comment]
Was having the same issue myself in 12.0.0, but using the correct command from both @JSoko and @mwalbeck has fixed it for now. Just to clarify, the command is:

`sudo -u www-data php /var/www/**NEXTCLOUD FOLDER**/occ maintenance:mimetype:update-db --repair-filecache`

**Comment by fthiery** [comment]
The workaround works (NC12), but needs to be repeated regulary; any idea why / how to fix this permanently ? Should i run this in a cron every minute ? Sounds like a bad solution to me...

**Comment by stratacast** [comment]
@fthiery did you solve your problem? The solution of putting the .json file in your nextcloud/config folder is a permanently lasting solution

**Comment by fthiery** [comment]
Yes it seems to work without the cron, sorry for not reporting.

**Comment by zeigerpuppy** [comment]
Thanks @JSoko and @mwalbeck! I can confirm the following works well with NC12 (no need to enable/disable app or run any crons):

```
nano /var/www/nextcloud/config/mimetypemapping.json
```

add kdbx mimetype
```
{
    "kdbx": ["x-application/kdbx"]
}

```

update database of mimetypes
```
sudo -u www-data php /var/www/nextcloud/occ maintenance:mimetype:update-db --repair-filecache
```

**Comment by e-alfred** [comment]
Maybe the useful information from @zeigerpuppy can be put into the main readme file of this repository? It would help people a lot to get Keeweb open the files in Nextcloud.

**Comment by ds4a** [comment]
I just wanted to feedback that the upgrade to Keeweb 0.4.0 under NC 11.0.2 has resolved this issue for me. Although just to note that once upgraded the app needs to disabled and then reenabled, just the once! Then all is good. I can now select a keeweb file from NC and Keeweb opens in NC as expected.

I Like the other improvements delivered in Keeweb 0.4.0.

Well done to the NC maintainers on these fixes and improvements. Much appreciated, this is one of the most useful apps on NC, especially when you have lots of details to remember.

**Comment by theScrabi** [comment]
I'm using Nextcloud 12.0.3 and Keeweb 0.4.0, however this error still occurs for me, and it seems [this](https://github.com/jhass/nextcloud-keeweb/issues/34#issuecomment-273595295) solution does not work for me either?

**Comment by KingPrawn22** [comment]
@theScrabi After updating to 12.0.3 myself, I found it wasn't working. Ran the following command ([as above](https://github.com/jhass/nextcloud-keeweb/issues/34#issuecomment-305520558)), and then after opening the database file from the file structure (as opposed to just launching Keeweb and expecting it to find it), it launches the Keeweb app. The app then recognises where the database is after this, and can be opened directly from the app. Worked for me anyhow and hope it does the same for you and any others on 12.0.3.

**Comment by theScrabi** [comment]
Yes I can confirm. That worked :)

**Comment by kevquirk** [comment]
I've not tested this workaround yet, but would updating the mimetype be required every time there is an NC update? Frustrating if that's the case. :(

**Comment by mammuth** [comment]
@kevquirk The webupdater does not override settings in the config :wink:

**Comment by kepuvv** [comment]
Thank, zeigerpuppy! Its work for me

**Comment by DMWOO7** [comment]
I can confirm that the mime-mapping from @zeigerpuppy is working fine on Nextcloud 12. Seems like that the command read the `mimetypemappings.json` file and copy all those mapping in the database. So afaik it should be possible for the nextcloud-keeweb extension to register it's mapping autmatically after installation?

If not, I agree that we should at least refer to this issue in the documentation/installation guide. Otherwise, a lot of people are wasting time to find out why this is not working. I also did cause it's confusing.

**Comment by ppmt** [comment]
My instance of NextCloud is on a hosted site so I don't have access to the command line. Is there another way to update the mime and database?

**Comment by ggeorgg** [comment]
Upload the following cmd.phpx file to your nextcloud root.
[cmd.zip](https://github.com/jhass/nextcloud-keeweb/files/1488586/cmd.zip)
I also had to comment out line 39 in console.php (I am on nextcloud 12.0.3 with php 7.1 cgi, no php 5.6.0)
Open your browser and navigate to the cmd.phpx you have recently uploaded (e.g. https://nextcloud.domain.com/cmd.phpx)
That should do it.
At least undo your changes. Delete the cmd.phpx again and comment in the line in console.php, otherwise your nextcloud instance will complain the next time you want to update)

**Comment by ppmt** [comment]
@ggeorg : thanks a lot for the script. I can now open my keepass file from Nextcloud!

I had to activate a module called posix to make the script work and rename it to php from phpx

**Comment by ludenticus** [comment]
Enabling/disabling the app has no effect on Nextcloud 12.0.3

**Comment by sunjam** [comment]
I've read through this thread and tried to setup Keeweb Webdav access on Nextcloud 12.0.4, but error I get is a little different.
![error-header](https://user-images.githubusercontent.com/1787238/33643919-abc89e24-d9f6-11e7-8292-f54be8c1556f.png)

**Comment by sbrl** [comment]
@sunjam I'd suggest opening a separate issue about that.

**Comment by sesipod** [comment]
I got it working on Nextcloud 12.0.4 - Debian 9 x64

I put my .kdbx on an " SMB / CIFS " share. 
Added the mimetypemapping.json file & restarted apache2.
[https://github.com/jhass/nextcloud-keeweb/issues/34#issuecomment-278634544](url)

I now can open the .kdbx form the SMB storage location.
Working for me on the following browsers....
Chrome `Version 64.0.3282.39 (Official Build) beta (32-bit)`
Firefox `57.0.3 (64-bit)`

**Comment by schnappijedi** [comment]
Wanted to add a workaround. Granted have never been able to edit a .kdbx file in Nextcloud but this will work for opening and otherwise using a database. Use a bookmark pointing to the Webdav location of one's .kdbx database such as: https://server/remote.php/webdav/Keepass-database.kdbx

Nextcloud has a bookmarking app that one could put this URL into or could keep in it a simple text document. Not ideal but works.

**Comment by zerpex** [comment]
Don't understand this issue, I got it working since Nextcloud 11 & 12 & 13.
many kdbx files on nextcloud file manager : a click on one of them open it directly on keypass... The expected behaviour, right ?

Using Nextcloud on docker with hoellen/nextcloud image : https://github.com/hoellen/dockerfiles/tree/master/nextcloud
z.

**Comment by DovendyrE** [comment]
Hi
I'm currently using keeweb in nextcloud 13.0.1
Just edited the /var/www/nextcloud/apps/keeweb/appinfo/info.xml to make it support nextcloud 13 then ran the command below to update mimetypes, and wokrs like charm here.
sudo -u www-data php /var/www/nextcloud/occ maintenance:mimetype:update-db --repair-filecache

**Comment by danger89** [comment]
So is it fixed or.. does it still require manual actions from the admin?

**Comment by Thatoo** [comment]
Would that plugin not help to solve the problem?
https://github.com/vavrecan/keeweb-local-server

**Comment by jhass** [comment]
Please update to 0.5.0. Pay attention to the required manual steps outlined in the readme.

I really whish there was a better way still...

**Comment by Mycah** [comment]
I'm having this issue in 16.0.4 with keeweb 0.5.1
I opened an issue with nextcloud
https://github.com/nextcloud/server/issues/17217

**Comment by pwnoobz** [comment]
I'm still having this issue on the last version of NextCloud and Debian9.

**Comment by kugiigi** [comment]
My problem here is that everytime the kdbx file is updated or it's newly uploaded, it's not detected correctly until I run file scan or update mimetype filecache. Does this really need to have a custom cron?
